rfcs/0136-scope-expression-registration.md

```diff
@@ -0,0 +1,129 @@
+
+1. We will add a new field to our service api definitions at all of the cluster, service, and endpoint levels
+   that allows us to define terms at their most appropriate level. No overriding will be permitted and each
+   term must be defined with both human language documentation and a regex defining valid values for parameters
+   submitted for these terms.
+1. Auth service will grow a `register()` endpoint that will allow a principal to upload structured documentation
+   of its own actions to the auth service. Each action will be prefixed by the name of the principal and will
+   require a scope to register actions in that namespace. For instance, only taskcluster-github will be allowed
+   to register the `github.createStatus` action. This action will have associated metadata including the scope
+   expression template that guards it. This registration must be versioned and have an expiration date.
+1. Auth service will grow an `authorize()` endpoint that will take as input an action, a clientId, and a set of
+   parameters that will be substituted into the action's scope expression template in order to turn it into a
+   scope expression. The endpoint will return a simple yes/no answer as to whether or not the clientId in question
+   has sufficient scopes to satisfy the scope expression with those parameters.
+1. Update all of our worker implementations (generic worker and script worker) to register their action somehow.
+   Each worker will be represented by a single action whose scope expression has many `if` statements for each
+   feature the worker supports. I expect that worker-manager will actually register these with auth to avoid
+   over-calling auth service. I think we can do this by making worker implementation part of the definition of a
+   workerType and having worker implementations generate the data that must be entered into the workerType definition.
+1. Build into the auth service a way to query all registered scope expression templates. The input will be a set of
+   scopes. The output will be a set of actions that may be authorized by these scopes and for each of these actions,
+   a list of possible values for parameters. This is a bit difficult to understand but the example should make this
+   more clear.
+
+# Example
+
+We create a new service called `taskcluster-dishwasher`. It washes dishes and has a single action called
+`dishwasher.wash`. This action is protected by a scope expression template:
+
+```json
+{
+  "AllOf": [
+    "dishwasher:wash:<detergent>"
+  ]
+}
+```
+
+The service will also define the term `detergent` with a bit of text about different values that might
+be present and also a regex that defines valid values for `detergent`.
+
+Before the service is started, when the cluster admin creates a client for this service, they will
+ensure that it has the scope `auth:register:dishwasher`. When the service is started, it will register
+this action and the associated terms with the auth service.
```

Comment:
I think we talked about a second set of scopes that allow access to specific scopes here (that's going to be confusing..).  So something like `auth:require-scope:dishwasher:*`.  I think in most cases these will be the same prefix (well, `dishwasher.*` and `dishwasher:*` here), but it might be more illustrative if they are different in the example.  Maybe the scope is `dishes:wash:<detergent>` instead?

In real life, I can see a case where multiple worker implementations use the same scope space for caches, for example, so worker manager would have `auth:register-scope:worker-cache:*`.